### PR TITLE
[FLINK-29668][Gelly] Remove Gelly

### PR DESCRIPTION
**Pull request by MartijnVisser**
## What is the purpose of the change

* Remove Gelly

## Brief change log

* Adds deprecation warning for the DataSet API to all DataSet API documentation, not just the overview page
* Removes all Gelly code
* Removes Gelly documentation
* Adds hint to check out Flink ML Iterations as a replacement for DataSet Iterations support in the DataSet deprecation warning

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Review by zentol — APPROVED** [review]
There's a gelly reference in the javadocs of `org.apache.flink.api.java.utils.DataSetUtils#checksumHashCode`.

File: flink-java/src/main/java/org/apache/flink/api/java/utils/DataSetUtils.java [hunk 1]
```diff
@@ -412,8 +412,7 @@ public TupleSummaryAggregator<R> reduce(
      *
      * @return A ChecksumHashCode that represents the count and checksum of elements in the data
      *     set.
-     * @deprecated replaced with {@code org.apache.flink.graph.asm.dataset.ChecksumHashCode} in
-     *     Gelly
+     * @deprecated replaced with {@code org.apache.flink.api.java.ChecksumHashCode}
```
Comment (zentol) [inline flink-java/src/main/java/org/apache/flink/api/java/utils/DataSetUtils.java:415]: `ChecksumHashCode` is not really a replacement.
Let's just say this method will simply be removed at some point.

**Review by tisonkun — APPROVED** [review]
Time past quickly :)

**Comment by MartijnVisser** [comment]
@flinkbot run azure

**Comment by MartijnVisser** [comment]
@flinkbot run azure

**Pull request merged**